pkg/networkservice/common/mechanisms/recvfd/server_test.go

```diff
@@ -0,0 +1,283 @@
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/edwarnicke/grpcfd"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/cls"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/common"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/kernel"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/peer"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/client"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/mechanisms/recvfd"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/mechanisms/sendfd"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+	"github.com/networkservicemesh/sdk/pkg/tools/sandbox"
+)
+
+type checkFileClosed struct {
+	onFileClosed map[string]func()
+
+	t *testing.T
+}
+
+type wrapTransceiver struct {
+	grpcfd.FDTransceiver
+	net.Addr
+
+	onFileClosed map[string]func()
+}
+
+func (n *checkFileClosed) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
+	return next.Server(ctx).Close(ctx, conn)
+}
+
+func (n *checkFileClosed) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		n.t.Fatal("No peer in context")
+	}
+
+	transceiver, ok := p.Addr.(grpcfd.FDTransceiver)
+	assert.True(n.t, ok)
+
+	p.Addr = &wrapTransceiver{
+		FDTransceiver: transceiver,
+		Addr:          p.Addr,
+		onFileClosed:  n.onFileClosed,
+	}
+
+	return next.Server(ctx).Request(ctx, request)
+}
+
+func (w *wrapTransceiver) RecvFileByURL(urlStr string) (<-chan *os.File, error) {
+	res, err := w.FDTransceiver.RecvFileByURL(urlStr)
+	if err != nil {
+		return nil, err
+	}
+
+	var chanRes = make(chan *os.File)
+	go func() {
+		for f := range res {
+			runtime.SetFinalizer(f, func(file *os.File) {
+				onFileClosedFunc, ok := w.onFileClosed[urlStr]
+				if ok {
+					onFileClosedFunc()
+				}
+			})
+			chanRes <- f
+		}
+	}()
+
+	return chanRes, nil
+}
+
+func createFile(fileName string, t *testing.T) (inodeURLStr string, fileClosedContext context.Context, cancelFunc func()) {
+	f, err := os.Create(fileName)
+	require.NoError(t, err, "Failed to create and open a file: %v", err)
+
+	info, err := f.Stat()
+	assert.NoError(t, err)
+
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	assert.True(t, ok)
+
+	err = f.Close()
+	require.NoError(t, err, "Failed to close file: %v", err)
+
+	fileClosedContext, cancelFunc = context.WithCancel(context.Background())
+	inodeURLStr = fmt.Sprintf("inode://%d/%d", stat.Dev, stat.Ino)
+
+	return
+}
```

Comment:
I do not quite follow why do we need this.

Reply:
This is required in case of multiple files to check that we're closing the same files that we created - I'm returning "inodeURLStr" to create a mapping `inodeURLStr -> onFileClosedFunc` which I'm later using in finalizer.